spec/Section 2 -- Source Schema.md

```diff
@@ -1,244 +1,399 @@
 }
 
-interface Node {
+interface Node @key(fields: "id") {
   id: ID!
 }
 ```
 
-When an entity resolver returns an interface all implementing types are inferred
-as entities.
+Lookup fields may return object, interface, or union types. In case a lookup
+field returns an abstract type (interface type or union type), all possible
+object types are considered entities and must have keys that correspond with the
+fields argument signature.
 
 ```graphql example
-extend type Query {
-  entityById(id: ID!, categoryId: Int): Entity @entityResolver
+type Query {
+  product(id: ID!, categoryId: Int): Product @lookup
+}
+
+union Product = Electronics | Clothing
+
+type Electronics @key(fields: "id categoryId") {
+  id: ID!
+  categoryId: Int
+  name: String
+  brand: String
+  price: Float
+}
+
+type Clothing @key(fields: "id categoryId") {
+  id: ID!
+  categoryId: Int
+  name: String
+  size: String
+  price: Float
+}
+```
+
+The following example shows an invalid lookup field as the `Clothing` type does
+not declare a key that corresponds with the lookup field's argument signature.
+
+```graphql counter-example
+type Query {
+  product(id: ID!, categoryId: Int): Product @lookup
 }
 
-union Entity = Cat | Dog
+union Product = Electronics | Clothing
 
-extend type Dog {
+type Electronics @key(fields: "id categoryId") {
   id: ID!
   categoryId: Int
+  name: String
+  brand: String
+  price: Float
 }
 
-extend type Cat {
+# Clothing does not have a key that corresponds
+# with the lookup field's argument signature.
+type Clothing @key(fields: "id") {
   id: ID!
   categoryId: Int
+  name: String
+  size: String
+  price: Float
+}
+```
+
+If the lookup returns an interface, the interface must also be annotated with a
+`@key` directive and declare it's keys.
+
+```graphql example
+interface Node @key(fields: "id") {
+  id: ID!
+}
+```
+
+Lookup fields must be accessible from the Query type. If not directly on the
+Query type, they must be accessible via fields that do not require arguments,
+starting from the Query root type.
+
+```graphql example
+type Query {
+  lookups: Lookups!
+}
+
+type Lookups {
+  productById(id: ID!): Product @lookup
+}
+
+type Product @key(fields: "id") {
+  id: ID!
+}
+```
+
+Lookups, can also be nested with other lookups and allow resolving nested
+entities that are part of an aggregate. In the following example the `Product`
+can be resolved by it's id but also the `ProductPrice` can be resolved by
+passing in a composite key containing the product id and region name of the
+product price.
+
+```graphql example
+type Query {
+  productById(id: ID!): Product @lookup
+}
+
+type Product @key(fields: "id") {
+  id: ID!
+  price(regionName: String!): ProductPrice @lookup
+}
+
+type ProductPrice @key(fields: "regionName product { id }") {
+  regionName: String!
+  product: Product
+  value: Float!
+}
+```
+
+Nested lookups must imediately follow the parent lookup and cannot be nested
+with fields inbetween.
```

Comment:
I believe we agreed that inbetween fields are allowed only if they don't have any arguments